### PR TITLE
Fixed #25731 -- Removed unused choices kwarg for Select.render()

### PR DESCRIPTION
**Pull request by jpic**
WIP for https://code.djangoproject.com/ticket/25731

### CODE REVIEW

**Comment by timgraham** [comment]
Looks like a good start. Merged the first commit in 65c13f9675d2ca7fc1c925e7182a2e35d07ff5fb. Can you fix the admin test failures?

**Comment by jpic** [comment]
Done, although I'm happy about the performance improvement, I'm a bit worried about it, is it possible that `chain(self.choices, choices)` would cause the QS in self.choices to be executed again ?

**Comment by charettes** [comment]
@jpic it should not be the case if queryset was already executed.

**Comment by jpic** [comment]
Windows still hates me , could we perhaps rerun the build ?

**Comment by timgraham** [comment]
Could please rebase your branch and squash commits? That will rereun the build but not the worry if there are problems. There's nothing here that's Windows specific.

**Comment by jpic** [comment]
Done, this seems like quite an important BC break though, shouldn't we do anything special about it ? Ie. engage the 2 version deprecation process and perhaps propose a way for app maintainers to support multiple versions ?

**Comment by timgraham** [comment]
As far as I know, this isn't part of any public API, however, the fact that an admin widget needed to be modified suggests that the change might cause problems for third-party widget authors and so a deprecation might be needed. Analyzing some third-party libraries to assess that would be useful.

**Comment by jpic** [comment]
I don't see any problem with the render() signature change, they should just stop passing it when calling super() on any django version since it wasn't used at all.

However, the arg order changes for render_options(), but it's not part of the public API to my knowledge. @codingjoe might want to throw in an opinion here because he's the ticket opener and also django-select2 which he maintains is affected https://github.com/applegrew/django-select2/blob/master/django_select2/forms.py#L231

DAL v3 is affected too, here's what it looks like

```
  File "/home/jpic/env/src/django/django/contrib/admin/widgets.py", line 288, in render
    'widget': self.widget.render(name, value, *args, **kwargs),
  File "/home/jpic/env/src/django/django/forms/widgets.py", line 522, in render
    options = self.render_options([value])
TypeError: render_options() takes exactly 3 arguments (2 given)
```

So, here's what it looks like for me:

```
    def render_options(self, *args):
        if django.VERSION < (1, 10):
            return self.future_render_options(args[1])
        return self.future_render_options(args[0])

    def super_future_render_options(self, selected_choices):
        if django.VERSION < (1, 10):
            return super(WidgetMixin, self).render_options(
                [], selected_choices)
        return super(WidgetMixin, self).render_options(
            selected_choices)

    def future_render_options(self, selected_choices):
        selected_choices = [c for c in selected_choices if c]

        if self.url and not selected_choices:
            self.choices = []

        return self.super_future_render_options(selected_choices)
```

Maybe we could add this to some django-compat lib, not sure which one is best out there.

Again, that doesn't look like public API so if that works for everybody it works for me.

**Comment by codingjoe** [comment]
@jpic I have this ticket on my radar. I will issue a new release of django-select2 once this is merged.
:+1: thanks for making this happening. Now I don't know what to do on djangoCon :wink:

**Comment by jpic** [comment]
Oops, well I probably found it while doing the same thing as you, we'll find something else for the sprint no worries B)

**Comment by timgraham** [comment]
As it seems at least a couple projects are using the private API, care to add some upgrade instructions in the backwards-incompatible changes section of the 1.10 release notes?

**Comment by jpic** [comment]
Done, to the same commit, not sure if i should have added a fixup commit, and wait to be asked for a rebase.

Once this is merged I'll try to get the snippet out there, hopefully in a django compat app, at least in a blog post or something. Or maybe should we just link to the comment that contains it in the release notes ?

**Comment by timgraham** [comment]
I'm not sure. If it's deserving of a mention, I tend to favor including it in the notes rather than link to it.

**Comment by jpic** [comment]
Maybe it's a bit too much for the release notes, let me know if there's anything else I can do to help. Thanks for your feedback !

File: docs/releases/1.10.txt [hunk 1]
```diff
@@ -206,6 +206,9 @@ Forms
 * Form and widget ``Media`` is now served using
   :mod:`django.contrib.staticfiles` if installed.
 
+* Select widget render_options() does not take the choices argument anymore,
```
Comment (timgraham) [inline docs/releases/1.10.txt]: This should be in "Backwards incompatible changes", not minor features. Is there a reason you mentioned `render_options()` but not `render()`?

**Comment by jpic** [comment]
I'm wondering if we could perhaps make the choices arg useful now by letting it override self.choices if set. This would allow a widget to render an arbitrary set of choices (ie. an autocomplete widget would only render selected choices). Perhaps @codingjoe you'd have an option on this ?

**Comment by timgraham** [comment]
merged in 926e90132dc15d76bb8d16e2f9f1279566cac3c3, thanks!

**Comment by codingjoe** [comment]
Since the argument position of `render_options` changed and this patch still calls it with a position it is impossible pro 3rd party maintainers to support the following example:

def render(*args, **kwargs):
    return self.render_choices(selected_choices=MY_CHOICES)
This breaks using both `django-autocomplete-light` or `django-select2`.

@timgraham what's your feeling about this? Should we patch it before the 1.10 release?

**Comment by codingjoe** [comment]
To be fair, I don't know if this is a common case, I changed it in the latest `django-select2` release. So far no one has complained and it doesn't effect Django itself anyways ;)

**Comment by timgraham** [comment]
I don't use these APIs myself, so I'll leave it to you and anyone else interested.

**Comment by jpic** [comment]
I don't know if it is "impossible", but if you know a way to improve this then go for it ;)

**Comment by codingjoe** [comment]
As of now, I don't see reason to act. My example will already break in the latest versions select2 and autocomplete-light, no matter which Django version. As long as no one complains, I presume there no one else out there exploiting this particular API.

**Comment by collinanderson** [comment]
(As a data point, the render_options() change broke my code too.)

**Comment by jpic** [comment]
Did it break any actual feature for you ? Or are you just talking about the
function signature change ?

**Pull request closed without merging**